File: crack-data-structures-and-algorithms/leetcode/3sum_closest_q16.py
# -*- coding: utf-8 -*-
# 0xCCCCCCCC

# Find three integers s.t the sum is closest to the target.
# Return the sum of these three integers.

# Given array nums = [-1, 2, 1, -4], and target = 1.
# The sum that is closest to the target is 2.

def three_sum_closest(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: int
    """
    # Like nomral 3-sum, but because we need to find the closest sum
    # thus we need bookkeeping and use offset to adjust scan direction.
    nums.sort()
    sum, dist = 0, 1 << 31
    i = 0
    while i < len(nums) - 2:
        l = i + 1
        r = len(nums) - 1
        while l < r:
            attempt = nums[i] + nums[l] + nums[r]
            offset = abs(attempt - target)
            # New closest found
            if offset < dist:
                dist = offset
                sum = attempt
            # If attempt < target, the sum is at the left of the target.
            # If attempt > target, the sum is at the right of the target.
            if attempt < target:
                l += 1
            elif attempt > target:
                r -= 1
            else:
                return sum
        i += 1
    return sum

nums = [-1, 2, 1, -4]
target = 1
print(three_sum_closest(nums, target))

File: crack-data-structures-and-algorithms/leetcode/3sum_closest_q16.cpp
#include <algorithm>
#include <vector>

using namespace std;

// 核心思路
// 和 3sum 类似，都是先排序然后分内外层处理
// 这里题目要求的是最接近目标和三数之和，所以固定好外层之后扫描内层，不断更新min_dist和closest
// 注意：可能出现sum和target相等的情况下，这种case直接返回
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        if (nums.size() < 3) {
            return 0;
        }

        sort(nums.begin(), nums.end());
        int closest = 0;
        int min_dist = INT_MAX;
        for (auto i = 0; i < nums.size() - 2; ++i) {
            auto l = i + 1;
            auto r = nums.size() - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                int diff = sum - target;
                if (diff == 0) {
                    return sum;
                }

                if (abs(diff) < min_dist) {
                    closest = sum;
                    min_dist = abs(diff);
                }

                if (diff > 0) {
                    --r;
                } else {
                    ++l;
                }
            }
        }

        return closest;
    }
};